File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height){
        int tempWater = 0;
        int totalWater = 0;
        int leftPointer = 0;
        int rightPointer = height.length-1;
        int leftMax = height[0];
        int rightMax = height[rightPointer];

        while (leftPointer < rightPointer){
            if (leftMax < rightMax){
                leftPointer ++;
                tempWater = leftMax - height[leftPointer];
                leftMax = Math.max(leftMax, height[leftPointer]);
            }else{
                rightPointer --;
                tempWater = rightMax - height[rightPointer];
                rightMax = Math.max(rightMax, height[rightPointer]);
            }

            if(tempWater > 0){
                totalWater += tempWater;
            }
        }
        return totalWater;
    }
}

File: LeetCode/Hard/0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        pref_max = [0] * n
        suf_max = [0] * n
        total_water_trapped = 0

        pref_max[0], suf_max[n-1] = height[0], height[n-1]

        for i in range(1, n):
            pref_max[i] = max(pref_max[i-1], height[i])

        for i in range(n-2, -1, -1):
            suf_max[i] = max(suf_max[i+1], height[i])

        for i in range(1, n-1):
            current_trapped_water = min(suf_max[i], pref_max[i]) - height[i]
            total_water_trapped += current_trapped_water
        
        return total_water_trapped